### PR TITLE
Add get_projects

### LINKED ISSUE #282: Projects missing get_projects() method
Please add get_projects() similar to other get_foo() methods in artifactory.py

@zhan9san

### PR DESCRIPTION
**Pull request by zhan9san**
Fix #282

### CODE REVIEW

**Comment by zhan9san** [comment]
I tried to consolidate the similar test functions into one. There would be many `if xxx` statement, and it's not easy for developer to identify the variable value in the context when error occurs.

So test cases for `Users`, `Groups` and `Projects` are split, and duplicate test code brings more convenient.

File: tests/unit/test_artifactory_path.py [hunk 1]
```diff
@@ -967,5 +971,190 @@ def test_from_aql_file(self):
...
+                "realm": "internal",
+            },
+        ]
+        user_1 = {"name": "user_1", "email": "user1@example.com"}
+        user_2 = {"name": "user_2", "email": "user2@example.com"}
+
+        # lazy = True
+        with responses.RequestsMock() as rsps:
+            rsps.add(responses.GET, request_url, json=users, status=200)
+            results = self.arti._get_all(lazy=True, url=url, key=key, cls=cls)
```
Comment (allburov) [inline tests/unit/test_artifactory_path.py:1002]: `_get_all` We should test only public methods

File: tests/unit/test_artifactory_path.py [hunk 2]
```diff
@@ -967,5 +971,190 @@ def test_from_aql_file(self):
         assert artifact.root == "/reponame/"
 
 
+class TestArtifactoryPathGetAll(unittest.TestCase):
+    # TODO: test repositories and permissions
+    def setUp(self):
+        self.arti = ArtifactoryPath("http://b.com/artifactory")
+
+    def test_get_all_users(self):
+        url = "/api/security/users"
```
Comment (allburov) [inline tests/unit/test_artifactory_path.py:980]: Can we move the preparation to a fixture? It's hard to get "what we actually do and what asserts are about" from long tests

File: tests/unit/test_artifactory_path.py [hunk 3]
```diff
@@ -967,5 +971,190 @@ def test_from_aql_file(self):
...
+        ]
+        user_1 = {"name": "user_1", "email": "user1@example.com"}
+        user_2 = {"name": "user_2", "email": "user2@example.com"}
+
+        # lazy = True
+        with responses.RequestsMock() as rsps:
+            rsps.add(responses.GET, request_url, json=users, status=200)
+            results = self.arti._get_all(lazy=True, url=url, key=key, cls=cls)
+
+            map(self.assertIsInstance, zip(results, repeat(cls)))
```
Comment (allburov) [inline tests/unit/test_artifactory_path.py:1004]: Does python3 call map right away? I thought it's a lazy function and you receive a map object

```python
>>> map(print, ['1', '2'])
<map object at 0x7f8811c92070>
>>> list(map(print, ['1', '2']))
1
2
[None, None]
>>> 
```

The one-liner is kinda tricky, let's use `for user in results` instead

File: tests/unit/test_artifactory_path.py [hunk 4]
```diff
@@ -967,5 +971,190 @@ def test_from_aql_file(self):
...
+            self.assertEqual(results[0].name, "user_1")
+            self.assertIsNone(results[0].email)
+            self.assertEqual(results[1].name, "user_2")
+            self.assertIsNone(results[1].email)
+
+            self.assertEqual(len(rsps.calls), 1)
+            self.assertEqual(rsps.calls[0].request.url, f"{request_url}")
```
Comment (allburov) [inline tests/unit/test_artifactory_path.py:1011]: Isn't f-string redundant here?

File: tests/unit/test_artifactory_path.py [hunk 5]
```diff
@@ -967,5 +971,190 @@ def test_from_aql_file(self):
...
+
+        # lazy = False
```
Comment (allburov) [inline tests/unit/test_artifactory_path.py:1013]: Let's create a separate test for lazy=true\false

```python
def test_get_all_users()
def test_get_all_users_lazy()
```

File: tests/unit/test_artifactory_path.py [hunk 6]
```diff
@@ -967,5 +971,190 @@ def test_from_aql_file(self):
...
+        user_2 = {"name": "user_2", "email": "user2@example.com"}
+
+        # lazy = True
+        with responses.RequestsMock() as rsps:
+            rsps.add(responses.GET, request_url, json=users, status=200)
+            results = self.arti._get_all(lazy=True, url=url, key=key, cls=cls)
+
+            map(self.assertIsInstance, zip(results, repeat(cls)))
+            self.assertEqual(results[0].name, "user_1")
+            self.assertIsNone(results[0].email)
```
Comment (allburov) [inline tests/unit/test_artifactory_path.py:1006]: The asserts about the instance should be inside the for loop too

**Comment by beliaev-maksim** [comment]
right now we do not have integration tests in CI/CD, only unittests are included

thus, please add some unittests

File: tests/unit/test_artifactory_path.py (hunk 1 above)
Comment (zhan9san) [reply tests/unit/test_artifactory_path.py]: Added `get_users`, `get_groups` and `get_projects` and removed `_get_all`.

File: tests/unit/test_artifactory_path.py (hunk 2 above)
Comment (zhan9san) [reply tests/unit/test_artifactory_path.py]: All fixtures are defined in `setUp`.

File: tests/unit/test_artifactory_path.py (hunk 3 above)
Comment (zhan9san) [reply tests/unit/test_artifactory_path.py]: Thank you for pointing this. This issue is fixed

File: tests/unit/test_artifactory_path.py (hunk 4 above)
Comment (zhan9san) [reply tests/unit/test_artifactory_path.py]: Fixed

File: tests/unit/test_artifactory_path.py (hunk 5 above)
Comment (zhan9san) [reply tests/unit/test_artifactory_path.py]: Fixed

File: tests/unit/test_artifactory_path.py (hunk 6 above)
Comment (zhan9san) [reply tests/unit/test_artifactory_path.py]: Fixed

**Comment by zhan9san** [comment]
The unit test shows the group url is different from groups'.

So I make a change in `group` endpoint as well.

Reference: [GetGroupDetails](https://www.jfrog.com/confluence/display/JFROG/Artifactory+REST+API#ArtifactoryRESTAPI-GetGroupDetails)

**Comment by beliaev-maksim** [comment]
@zhan9san 
looks great!

File: tests/unit/test_artifactory_path.py [hunk 7]
```diff
@@ -1158,5 +1161,229 @@ def test_from_aql_file(self):
...
+            )
+
+            results = self.arti.get_projects(lazy=True)
+
+            for project in results:
+                self.assertIsInstance(project, Project)
+            self.assertEqual(results[0].project_key, "project_key_1")
+            self.assertNotEqual(results[0].description, "description_1")
+            self.assertEqual(results[1].project_key, "project_key_2")
+            self.assertNotEqual(results[1].description, "description_2")
```
Comment (beliaev-maksim) [inline tests/unit/test_artifactory_path.py:1382]: why do we do a not Equal?
may be worth to check that it is kept as default class value? eg according to [L1315](https://github.com/devopshq/artifactory/blob/a14e7bf639c0a389bf104bdd9f25d32715113f13/dohq_artifactory/admin.py#L1315)
```suggestion
            self.assertFalse(results[1].description)
```

Comment (zhan9san) [reply tests/unit/test_artifactory_path.py:1382]: how about self.assertEqual(results[1].description, "")?

Comment (beliaev-maksim) [reply tests/unit/test_artifactory_path.py:1382]: usually in the code you will run `if not var.description`, but for this particular case to check that bool or other value was not leaked that makes sense
go ahead

Comment (zhan9san) [reply tests/unit/test_artifactory_path.py:1382]: Thanks. Will do it tomorrow

Comment (zhan9san) [reply tests/unit/test_artifactory_path.py]: Hi @beliaev-maksim 

This issue is fixed. Could you kindly review and approve this PR?

**Pull request merged**